CSI MountDevice/UnmountDevice Implementation

davidz627 opened the pull request:
Fixes #60114

**What this PR does / why we need it**:
This PR Implements MountDevice and UnmountDevice for the CSI Plugin, the functions will call through to NodeStageVolume/NodeUnstageVolume for CSI plugins.

/sig storage

```release-note
Implements MountDevice and UnmountDevice for the CSI Plugin, the functions will call through to NodeStageVolume/NodeUnstageVolume for CSI plugins.
```

davidz627 committed:
Added MountDevice/UnmountDevice pass-through to NodeStageVolume/NodeUnstageVolume for CSI Volume Plugin. Added related unit tests. Vendored CSI Spec to HEAD

davidz627 commented:
/assign saad-ali

davidz627 commented:
/assign vladimirvivien

saad-ali commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -229,12 +235,167 @@ func (c *csiAttacher) VolumesAreAttached(specs []*volume.Spec, nodeName types.No
+	glog.V(4).Infof(log("attacher.MountDevice(%s, %s)", devicePath, deviceMountPath))
+
+	// Setup
+	ctx, cancel := grpctx.WithTimeout(grpctx.Background(), csiTimeout)
```
Move creating `ctx` down to right before you actually use it for the first time. No need to waste cycles if you end up not using it.

saad-ali commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -229,12 +235,167 @@ func (c *csiAttacher) VolumesAreAttached(specs []*volume.Spec, nodeName types.No
+		return fmt.Errorf("attacher.MountDevice failed while getting node capabilities: %v", err)
+	}
+
+	if capabilities == nil {
```
Move this nil check inside `hasStageUnstageCapability(...)`

saad-ali commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -229,12 +235,167 @@ func (c *csiAttacher) VolumesAreAttached(specs []*volume.Spec, nodeName types.No
+		return fmt.Errorf("attacher.MountDevice failed, deviceMountPath is empty")
+	}
+
+	mounted, err := isDirMounted(c.plugin, deviceMountPath)
```
Could you put this check first, and if it is false, then create a context, client, etc.?

saad-ali commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -229,12 +235,167 @@ func (c *csiAttacher) VolumesAreAttached(specs []*volume.Spec, nodeName types.No
+	}
+
+	nodeName := string(c.plugin.host.GetNodeName())
+	fmt.Printf("voldid: %v, drivername: %v, nodeName: %v", csiSource.VolumeHandle, csiSource.Driver, nodeName)
```
Remember to remove debug statement or convert to glog with appropriate verbosity

saad-ali commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -229,12 +235,167 @@ func (c *csiAttacher) VolumesAreAttached(specs []*volume.Spec, nodeName types.No
+
+	// probe driver
+	// TODO (vladimirvivien) move probe call where it is done only when it is needed.
+	if err := csi.NodeProbe(ctx, csiVersion); err != nil {
```
You can remove probe call altogether. It has been removed in the latest (HEAD) version of CSI.

saad-ali commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -229,12 +235,167 @@ func (c *csiAttacher) VolumesAreAttached(specs []*volume.Spec, nodeName types.No
+	publishVolumeInfo := attachment.Status.AttachmentMetadata
+
+	// get volume attributes
+	// TODO: for alpha vol attributes are passed via PV.Annotations
```
Make sure to do a follow up PR to fix this, create a bug or something to remind you.

saad-ali commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -229,12 +235,167 @@ func (c *csiAttacher) VolumesAreAttached(specs []*volume.Spec, nodeName types.No
+		return "", fmt.Errorf("attacher.GetDeviceMountPath failed, driver name empty")
+	}
+
+	deviceMountPath := path.Join(c.plugin.host.GetPluginDir(csiSource.Driver), "mounts", csiSource.VolumeHandle)
```
And maybe sanitize the volume and driver names (`EscapeQualifiedNameForDisk` or something like that)?

saad-ali commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -335,10 +496,80 @@ func (c *csiAttacher) waitForVolumeDetachmentInternal(volumeHandle, attachID str
+	return stageUnstageSet, nil
+}
+
+func getDriverAndVolNameFromDeviceMountPath(deviceMountPath string, plugin *csiPlugin) (string, string, error) {
```
Oh man this is a fun bit of code.

vladimirvivien replied:
It would be helpful to document the format deviceMountPath (i.e. `{pluginDir}/csi/etc` so future maintainers would have an idea why this works.

saad-ali commented on pkg/kubelet/kubelet_getters.go:
```diff
@@ -64,6 +64,12 @@ func (kl *Kubelet) getPluginDir(pluginName string) string {
+// getPluginDir returns the pluginName for the given pluginDir that
+// was created by getPluginDir
+func (kl *Kubelet) getPluginNameFromDir(pluginDir string) string {
+	return filepath.Base(pluginDir)
```
Oh it looks like `getDriverAndVolNameFromDeviceMountPath(...)` is called first that splits at `/` boundaries, and then `getPluginNameFromDir(...)` is called to finish everything off.

Why not just do `GetPluginDir(...)` inside `getPluginNameFromDir(...)` and move this logic there? That way you could eliminate this method all together. It doesn't look like it is used anywhere else.

vladimirvivien commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -335,10 +496,80 @@ func (c *csiAttacher) waitForVolumeDetachmentInternal(volumeHandle, attachID str
+	}
+	volID := strings.Join(splitted[len(splitted)-3:], "/")
+	dir := strings.Join(splitted[:len(splitted)-4], "/")
+	pluginName := plugin.host.GetPluginNameFromDir(dir)
```
Assuming this`{pluginDir}/csi/{csiDriverName}/mounts/{csiVolumeHandle}` is the format, then  
`strings.Join(splitted[len(splitted)-3:], "/")` will return slice like `{csiDriverName}/mounts/{csiVolumeHandle}` .  Uneasy about the hardcoded offsets, seems a bit brittle.

Also, consider using filepath.Base(), filepath.Dir(), filepath.Join() for more portable path nomenclatures.

davidz627 replied:
I agree that the hardcoded offesets are brittle but I couldn't find a better way to do this without a hardcoded number of "Dir"'s anyway (3 of them). Because {csiVolumeHandle} itself is a 3 part object of form `project/zone/name`. That fact isn't really captured anywhere (maybe I should have a comment). Open to suggestions for better ways to do this.

I tried using `Rel` which ended up being more brittle because even small changes or differences to directory structure would completely destroy the resulting path.

vladimirvivien reviewed (commented):
Please take a look at comments.

davidz627 commented:
@saad-ali @vladimirvivien Addressed all comments. Biggest change is in `makeDeviceMountPath` and `getDriverAndVolNameFromDeviceMountPath`. Please take a closer look there. Thanks!

davidz627 commented:
@vladimirvivien the PV method in `getDriverAndVolNameFromDeviceMountPath` was created to be able to reliably retrieve the volume handle.

Since there are no character restrictions in the CSI spec on volume handles there were no guaranteed ways to parse the volume handle out of the DeviceMountPath even with sanitation.

davidz627 commented:
/retest

davidz627 commented:
/retest

saad-ali commented on pkg/volume/csi/csi_client.go:
```diff
@@ -158,6 +173,82 @@ func (c *csiDriverClient) NodeUnpublishVolume(ctx grpctx.Context, volID string,
+		return err
+	}
+
+	req := &csipb.NodeStageVolumeRequest{
```
Pass in secret (`node_stage_secrets` field). It is specified on the Kubernetes side in the `NodeStageSecretRef` in `CSIPersistentVolumeSource`.

Examples here:
https://github.com/kubernetes/kubernetes/pull/60118
https://github.com/kubernetes/kubernetes/pull/60382

saad-ali commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -229,12 +241,125 @@ func (c *csiAttacher) VolumesAreAttached(specs []*volume.Spec, nodeName types.No
+		return fmt.Errorf("attacher.MountDevice failed while getting node capabilities: %v", err)
+	}
+
+	stageUnstageSet, err := hasStageUnstageCapability(capabilities)
```
Since you always call `NodeGetCapabilities(...)` before `hasStageUnstageCapability(...)`. Maybe move `NodeGetCapabilities(...)` inside `hasStageUnstageCapability(...)`

saad-ali reviewed (commented):
Almost there!

vladimirvivien commented:
@davidz627 taking a look right now.

davidz627 commented:
/retest

saad-ali reviewed (approved):
/lgtm
/approve

vladimirvivien commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -335,12 +460,111 @@ func (c *csiAttacher) waitForVolumeDetachmentInternal(volumeHandle, attachID str
+	if file := filepath.Base(deviceMountPath); file != globalMountInGlobalPath {
+		return "", "", fmt.Errorf("getDriverAndVolNameFromDeviceMountPath failed, path did not end in %s", globalMountInGlobalPath)
+	}
+	// dir is now /var/lib/kubelet/plugins/kubernetes.io/csi/pv/{pvname}
```
Is `{pvname}` sanitized elsewhere (to remove path demarkation chars) ?  If so, it probably need to return back to its original state.  If no sanitization occurred, ignore.

davidz627 replied:
I am not doing sanitation anywhere, so the name should be same in both the creation and the extraction.

vladimirvivien commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -335,12 +460,111 @@ func (c *csiAttacher) waitForVolumeDetachmentInternal(volumeHandle, attachID str
+	pvName := filepath.Base(dir)
+
+	// Get PV and check for errors
+	pv, err := k8s.CoreV1().PersistentVolumes().Get(pvName, meta.GetOptions{})
```
Do we need to worry about `namespace` here?  If not, there maybe name collision.

davidz627 replied:
PV's are not namespaced in Kubernetes.

vladimirvivien commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -335,12 +460,111 @@ func (c *csiAttacher) waitForVolumeDetachmentInternal(volumeHandle, attachID str
+	}
+
+	// Get VolumeHandle and PluginName from pv
+	csiSource := pv.Spec.CSI
```
Do a nil check here for `pv.Spec.CSI`

davidz627 replied:
Done 5 lines above already

vladimirvivien reviewed (commented):
Added some comments.  Getting there.

davidz627 commented:
/retest

saad-ali commented:
Will let @vladimirvivien review and LGTM

/lgtm cancel

saad-ali commented:
/approve

davidz627 commented:
/retest

davidz627 commented:
/retest

davidz627 commented:
/retest

vladimirvivien commented:
/LGTM
